### PR TITLE
docs & small optimizations to dom ops

### PR DESCRIPTION
**Pull request by richrobber2**
i added some js docs and made some dom optimizations such as:
- switching to using Object.assign for some variable assignments to avoid multiple dom calls
- using guard clauses 
- used optional chaining operator to avoid errors
- and more but you gotta look over the code changes for more

### CODE REVIEW

**Comment by richrobber2** [comment]
any more speed ups would have to come from the .svlte file for the gradio component

**Comment by daswer123** [comment]
Cool, the documentation will definitely help others understand what's even going on in the code :)
And general improvement of the code is also good.

I'll go through the code now, test it on different versions and I think it will be ready to merge :)

File: javascript/canvas-zoom.js [hunk 1]
```diff
@@ -362,20 +362,23 @@
     }
 
     const isGetSizeImgBtnExists = !!document.querySelector("#img2img_detect_image_size_btn");
-
+    let getImgDataBtn, clonedDiv;
```
Comment (daswer123) [inline javascript/canvas-zoom.js:365]: Bug fix for version 1.2 and lower

File: javascript/canvas-zoom.js [hunk 2]
```diff
@@ -770,45 +773,48 @@
...
-        if (canvasOpacity > 1 && getActiveTab(elements).innerText === "Inpaint") {
-          setCanvasOpacity(1);
-          setBrushOpacity(1);
-          canvasOpacity = brushOpacity = 1;
-        }
+     /**
+     * Restores the source of an image to its original state, in case of an undo operation.
+     */
+    //Restore undo func
+    function restoreUndo() {
```
Comment (daswer123) [inline javascript/canvas-zoom.js:780]: Didn't get a chance to save your solution, but I optimized this one

File: javascript/canvas-zoom.js [hunk 3]
```diff
@@ -823,20 +829,23 @@
...
-          `${activeElement} .brush`.parentElement,
-          `${activeElement} .canvas-tooltip-info--extension`
-        ];
-
-        selectors.forEach(selector => {
-          const element = document.querySelector(selector);
-
-          element.classList.toggle("canvas-buttons-hidden", hide);
-          element.classList.toggle("canvas-buttons-visible", !hide);
+      function hideCanvasButtons(hide = false) {
```
Comment (daswer123) [inline javascript/canvas-zoom.js:832]: I couldn't save your solution here either, but I optimized what I had and it seems fine

File: javascript/canvas-zoom.js [hunk 4]
```diff
@@ -955,9 +964,13 @@
 
         const shouldHideButtons = window.isDrawing && hotkeysConfig.canvas_zoom_hide_btn;
 
-        if (shouldHideButtons !== buttonsHidden) {
-          hideCanvasButtons(shouldHideButtons);
-          buttonsHidden = shouldHideButtons;
+        try {
```
Comment (daswer123) [inline javascript/canvas-zoom.js:967]: For some reason there was a bug, applied a small hack that would not have errors

**Comment by daswer123** [comment]
In general checked on all versions from March 25 to version 1.5, everything works, no errors noticed. I think we can do a merge, a little more testing and that's it :)

**Comment by richrobber2** [comment]
yes, do note that i didnt do a ton of testing

**Comment by daswer123** [comment]
Checked all versions again, everything works.

**Pull request merged**